### DISCUSSION TITLE
RPi Camera + USB microphone

### DISCUSSION THREAD

**Discussion by alegag** [bluenviron/mediamtx#1388] [Questions]
Hi,

  I'm trying to merge a video stream with an audio stream.

  The Camera is a Raspberry Pi Camera Module V3.  I'm able to make it stream by itself.

  And the audio source is from a USB microphone, which I'm able to record from with `ffmpeg`.

  To record from the USB microphone:

```bash
ffmpeg -ar 8000  -ac 1 -f alsa -i hw:1 -ac 1 -acodec aac -b:a 512k -f mp4 test.mp4
```

  This works properly.

  In the rtsp-simple-server configuration, I have this for the camera:

```yml
paths:
  cam:
    source: rpiCamera
    rpiCameraWidth: 1920
    rpiCameraHeight: 1080
```

  Which also works properly.

  Now, based on other question, I have tried to merge using `runOnDemand` and `ffmpeg`, but the stream crash when I request it.

  Below the section above, I have added this:

```yml
  audio:
    runOnDemand: ffmpeg -hide_banner -loglevel error -i rtsp://localhost:8554/cam -c:v copy -f alsa -ac 1 -i hw:1 -af "volume=20dB" -acodec aac -b:a 512k -f rtsp rtsp://localhost:8554/audio
```

  Any idea what I'm missing here ?

  If I remove the `-i rtsp://localhost:8554/cam -c:v copy` part, I'm able to stream the audio only, so the problem seems to be with getting the camera stream.

Thank you !

**Comment by aler9** [comment]
Hello, this is a FFmpeg question, not a server question, anyway, from my experience, input streams must be put at the beginning of the command, then you have to map them, then you can choose the codec, and finally you can send out the stream.

Example:

```
ffmpeg -hide_banner -loglevel error \
-i rtsp://localhost:8554/cam \
-i hw:1 \
-map 0:v:0 \
-map 1:a:0 \
-c:v copy -f alsa -ac 1 -af "volume=20dB" -acodec aac -b:a 512k \
-f rtsp rtsp://localhost:8554/audio
```

**Comment by aler9** [comment]
By the way, the README has been updated with step-by-step instructions on how to add audio to an existing RPI Camera stream. GStreamer has been chosen over FFmpeg because tests showed that it introduces less latency:

https://github.com/bluenviron/mediamtx#from-a-raspberry-pi-camera

**Reply by ichalmers** [reply]
I've used your example to get this running on a rpi zero with a microphone.  The one thing I need to do is change the audio from opus to aac.  I've been trying to change the gstreamer settings to make this work (without luck).

Is there any chance of an example which would output aac?

**Reply by aler9** [reply]
> Is there any chance of an example which would output aac?

Replace `opusenc` with `avenc_aac`, which is documented [here](https://gstreamer.freedesktop.org/documentation/libav/avenc_aac.html?gi-language=c).

[a comment by an automated account is not shown]
